### PR TITLE
Make switch in nfo parser smaller

### PR DESCRIPTION
**Pull request by daullmer**
**Changes**

- Use custom methods to read strings, ints, floats, bools and dates from nfo
- Move subtree reader methods to separate file

The best way to review to PR would probably be to go through each commit which may make it easier to understand.

**Issues**
Fix multiple language tags resulting in only the last country to be added to the database

### CODE REVIEW

**Review by crobibero — APPROVED** [review]
Much clean

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs [hunk 1]
```diff
@@ -0,0 +1,246 @@
...
+            return value;
+        }
+
+        internal static void ReadProviderIdFromNfo(this XmlReader xmlReader, IHasProviderIds item, Dictionary<string, string> providerIds)
+        {
+            string readerName = xmlReader.Name;
+            if (providerIds.TryGetValue(readerName, out string? providerIdValue))
+            {
+                var id = xmlReader.ReadElementContentAsString();
+                if (!string.IsNullOrWhiteSpace(providerIdValue) && !string.IsNullOrWhiteSpace(id))
```
Comment (BaronGreenback) [inline MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:65]: Just checking, but with providerIds.TryGetValue, is string.IsNullOrWhiteSpace(providerIdValue) required?

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs [hunk 2]
```diff
@@ -0,0 +1,246 @@
...
+                reader.Read();
+                return;
+            }
+
+            var provider = reader.GetAttribute("type");
+            var id = reader.ReadElementContentAsString();
+            if (!string.IsNullOrWhiteSpace(provider) && !string.IsNullOrWhiteSpace(id))
+            {
+                item.SetProviderId(provider, id);
+            }
```
Comment (BaronGreenback) [inline MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:116]: I take it from this you are parsing nicely and ignoring errors in the nfo? 

Just thinking about debugging / tracing issues - would it help to have some sort of logging?

File: MediaBrowser.XbmcMetadata/Parsers/NfoSubtreeParsers.cs [hunk 3]
```diff
@@ -0,0 +1,538 @@
...
+                            var val = reader.ReadElementContentAsString();
+
+                            if (!string.IsNullOrWhiteSpace(val))
+                            {
+                                if (float.TryParse(val, NumberStyles.AllowDecimalPoint, CultureInfo.InvariantCulture, out var ratingValue))
+                                {
+                                    // if ratingName contains tomato --> assume critic rating
+                                    if (ratingName != null &&
+                                        ratingName.Contains("tomato", StringComparison.OrdinalIgnoreCase) &&
+                                        !ratingName.Contains("audience", StringComparison.OrdinalIgnoreCase))
```
Comment (BaronGreenback) [inline MediaBrowser.XbmcMetadata/Parsers/NfoSubtreeParsers.cs:514]: Can you have both "tomato" and "audience" in the ratingName ?
```suggestion
                                        !ratingName.Contains("audience", StringComparison.OrdinalIgnoreCase))
```

**Review by BaronGreenback — APPROVED** [review]

File: MediaBrowser.XbmcMetadata/Parsers/NfoSubtreeParsers.cs (hunk 3 above)
Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/NfoSubtreeParsers.cs:514]: Yes there could be a "tomatometerallaudience" rating https://github.com/jellyfin/jellyfin/pull/4981#discussion_r554436654

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs (hunk 1 above)
Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:65]: Probably not. I just copied most of the code from the Parser class and didn't give all of it a closer look

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs (hunk 2 above)
Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:116]: Yes I thought about adding logging if the value is invalid. Should it be a warning, info, or debug log?

Comment (BaronGreenback) [reply MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:116]: i'd personally do debugs.

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs (hunk 1 above)
Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:65]: Just checked, `providerIdValue` contains at least "Id", so I removed the check

File: MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+using System.Linq;
+using System.Xml;
+using MediaBrowser.Controller.Entities;
+using MediaBrowser.Model.Entities;
+using MediaBrowser.XbmcMetadata.Savers;
+using Microsoft.Extensions.Logging;
+
+namespace MediaBrowser.XbmcMetadata.Parsers
+{
+    internal class NfoParserHelpers
```
Comment (cvium) [inline MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:13]: There's no state in this class. New'ing it up every time you need to parse XML is inefficient. It should be registered as a singleton or remove the logger and make it completely static.

Comment (Ullmie02) [reply MediaBrowser.XbmcMetadata/Parsers/NfoParserHelpers.cs:13]: The parser itself is also recreated every time a nfo file is read. So would it be best if I made that a singleton as well?

**Pull request closed without merging**